feat: unified registry

hiddentao opened the pull request:
Related to #19

Arachnid commented on contracts/src/registry/PermissionedRegistry.sol:
```diff
@@ -2,55 +2,264 @@
+
+    uint256 public constant ROLE_RENEW = 1 << 1;
+    uint256 public constant ROLE_RENEW_ADMIN = ROLE_RENEW << 128;
+
+    uint256 public constant ROLE_SET_SUBREGISTRY = 1 << 2;
+    uint256 public constant ROLE_SET_SUBREGISTRY_ADMIN = ROLE_SET_SUBREGISTRY << 128;
 
-abstract contract PermissionedRegistry is BaseRegistry {
-    uint96 public constant FLAGS_MASK = 0x7;
-    uint96 public constant FLAG_SUBREGISTRY_LOCKED = 0x1;
-    uint96 public constant FLAG_RESOLVER_LOCKED = 0x2;
-    uint96 public constant FLAG_FLAGS_LOCKED = 0x4;
+    uint256 public constant ROLE_SET_RESOLVER = 1 << 3;
+    uint256 public constant ROLE_SET_RESOLVER_ADMIN = ROLE_SET_RESOLVER << 128;
+
+    uint256 public constant ROLE_SET_FLAGS = 1 << 4;
+    uint256 public constant ROLE_SET_FLAGS_ADMIN = ROLE_SET_FLAGS << 128;
+
+    uint256 public constant MAX_EXPIRY = type(uint64).max;
+    uint96 public constant FLAGS_MASK = 0xffffffff; // 32 bits
```
I'm a bit concerned about the bytecode and runtime overhead of all these public constants. What do you think about making them private?

hiddentao replied:
That's fine, but we then have to accept front-end brittleness, i.e things could go wrong.

Arachnid commented on contracts/src/registry/PermissionedRegistry.sol:
```diff
@@ -2,55 +2,264 @@
+        datastore.setSubregistry(tokenId, address(registry), flags);
+        datastore.setResolver(tokenId, resolver, flags);
+
+        if (bytes(_uri).length > 0) {
+            metadataProvider.setTokenUri(tokenId, _uri);
+            emit URI(_uri, tokenId);
+        }
```
Aside from the security issue, I don't think we need per-token URIs in general. An interface that takes the token ID and generates it based on an algorithm should work fine.

hiddentao replied:
Well we have support for any scheme. That's the point the Registry metadata implementations. Are you saying to assume we'll never externally set the uri per tokenId?

Arachnid replied:
I can't see a need to set per-token URIs in the root, .eth, or user registries. So I think for now we can leave them out. A subclass of this could always introduce them if needed.

Arachnid commented on contracts/src/registry/PermissionedRegistry.sol:
```diff
@@ -2,55 +2,264 @@
+        return tokenId;
+    }
+
+    function setTokenObserver(uint256 tokenId, address _observer) external onlyTokenOwner(tokenId) {
```
Should there be a permission to set a token observer?

hiddentao replied:
Well, only token owner can set it. That should do, right?

Arachnid replied:
Right, I'm suggesting it would be more flexible if that's a right that can be delegated like others here.

Arachnid commented on contracts/src/registry/PermissionedRegistry.sol:
```diff
@@ -2,55 +2,264 @@
+        emit TokenObserverSet(tokenId, _observer);
+    }
+
+    function renew(uint256 tokenId, uint64 expires) public onlyRootRoles(ROLE_RENEW) {
```
Perhaps this should be `onlyRoles` instead of `onlyRootRoles`; it could be useful to grant a user the right to renew their own name (cost-free, for user registries for example). We'd need to mask this bit out in the .eth registrar.

Arachnid commented on contracts/src/registry/PermissionedRegistry.sol:
```diff
@@ -2,55 +2,264 @@
     }
 
-    function flags(uint256 tokenId) external view returns(uint96) {
+    function setFlags(uint256 tokenId, uint96 _flags)
+        external
+        onlyRoles(tokenIdResource(tokenId), ROLE_SET_FLAGS)
+        returns (uint256 newTokenId)
+    {
+        (address subregistry, uint96 oldFlags) = datastore.getSubregistry(tokenId);
+        uint96 newFlags = oldFlags | (_flags & FLAGS_MASK);
+
+        if (newFlags != oldFlags) {
+            datastore.setSubregistry(tokenId, subregistry, newFlags);
+
+            newTokenId = (tokenId & ~uint256(FLAGS_MASK)) | (newFlags & FLAGS_MASK);
+            if (tokenId != newTokenId) {
+                address owner = ownerOf(tokenId);
+                _mint(owner, newTokenId, 1, "");
+                _burn(owner, tokenId, 1);
+            }
+        } else {
+            newTokenId = tokenId;
+        }
+    }
```
Are we 100% sure we still need flags, now that we have the new permission model?

hiddentao replied:
Well, I wasn't sure what else you guys wanted flags for so left them in...should I remove them? This would also mean being able to cleanup the "token id changes when flags change" logic, right?

Arachnid replied:
If we're not using them now, then yes we can.

The issue the token ID change was created to fix is present here in another form - we don't want a situation where someone can grant themselves permissions just prior to transferring the domain to the new owner, and we need a way to make sure the user has 'clean title' to the name they just bought. Needs some thought on how to handle this.

hiddentao replied:
See https://github.com/ensdomains/namechain/pull/29#issuecomment-2760497409

Arachnid commented on contracts/src/registry/PermissionedRegistry.sol:
There's a vulnerability here whereby a name that expires and is reregistered still has all the old name's ACLs. Likewise, we need to provide a way to invalidate ACLs when a name is transferred to a new owner.

One way to do this is with a 'version' field, but it's an extra element to store. We could mask off bits in the token ID to store this. Thoughts?

hiddentao replied:
Sounds like we need a "Revoke all roles for all users" thing, like I mentioned before :)

Arachnid replied:
Right, but that needs to be compatible with also being able to permanently revoke permissions on a name. Any thoughts on how to achieve both objectives?

hiddentao replied:
See https://github.com/ensdomains/namechain/pull/29#issuecomment-2760497409

hiddentao commented:
@Arachnid 

i've removed flags completely. Note that `RegistryDatastore` now just stores the expiry and address in the `uint256`, leaving the leftover bits doing nothing for now.

As for solving the two issues you mentioned (people assigning roles to themselves prior to a transfer / expired names still have old role assignments when being registered again), i've implemented a versioning system in the tokenIdResource() method, so it's the `PermissionedRegistry` level. A benefit of this is that the NFT token id now doesn't need to change, bringing stability across the front-end side of things - i.e. once a name is created the NFT token id for it never changes.

Arachnid commented on contracts/src/registry/PermissionedRegistry.sol:
```diff
@@ -2,55 +2,215 @@
+/**
+ * @dev Observer pattern for events on existing tokens.
+ */
+interface PermissionedRegistryTokenObserver {
```
Can probably just call this `TokenObserver`.

Arachnid commented on contracts/src/registry/BaseUriRegistryMetadata.sol:
```diff
@@ -22,6 +23,6 @@ contract BaseUriRegistryMetadata is AccessControl, IRegistryMetadata {
 
     function supportsInterface(bytes4 interfaceId) public view override returns (bool) {
-        return interfaceId == type(IRegistryMetadata).interfaceId || super.supportsInterface(interfaceId);
+        return interfaceId == type(IRegistryMetadata).interfaceId || EnhancedAccessControl.supportsInterface(interfaceId);
```
The original implementation here is correct.

Arachnid commented on contracts/src/registry/ETHRegistrar.sol:
```diff
@@ -22,16 +27,17 @@ contract ETHRegistrar is IETHRegistrar, AccessControl {
     error InsufficientValue(uint256 required, uint256 provided);
 
-    IETHRegistry public immutable registry;
+    IPermissionedRegistry public immutable registry;
```
We should probably define an interface specific to a registry-with-register-and-renew functions that's more general than IPermissionedRegistry.

hiddentao replied:
I need more guidance here. Should it have relinquish as well? what about the getters to fetch expiry, etc? and that setters for subregistry and resolver value?

Arachnid replied:
Ideally it should be a useful interface that's a superset of `IRegistry`, but not specific to `PermissionedRegistry`. So it would need to include `register`, `renew` and any other methods this contract uses, but needn't include methods this contract doesn't use unless they're naturally part of the same interface. So the getters for expiry etc would naturally be included because their setters are, for example.

hiddentao replied:
Ok, I've just updated `IRegistry` directly

Arachnid commented on contracts/src/registry/EnhancedAccessControl.sol:
```diff
@@ -188,23 +188,24 @@ abstract contract EnhancedAccessControl is Context, ERC165 {
      * @dev Grants multiple roles to `account`.
      */
-    function _grantRoles(bytes32 resource, uint256 roleBitmap, address account) internal virtual returns (bool) {
+    function _grantRoles(bytes32 resource, uint256 roleBitmap, address account, bool enableCallbacks) internal virtual returns (bool) {
```
Perhaps `executeCallbacks` instead of `enableCallbacks`; 'enable' implies a persistent change.

Arachnid commented on contracts/src/registry/PermissionedRegistry.sol:
```diff
@@ -2,55 +2,269 @@
-        withSubregistryFlags(tokenId, FLAG_FLAGS_LOCKED, 0)
-        returns(uint96 newFlags)
+    constructor(IRegistryDatastore _datastore, IRegistryMetadata _metadata) BaseRegistry(_datastore) MetadataMixin(_metadata) {
+        _grantRoles(ROOT_RESOURCE, ALL_ROLES, _msgSender(), false);
```
When instantiating this in a trustless context - for example, the .eth registry - we need to revoke or not set `ROLE_SET_SUBREGISTRY` and `ROLE_SET_RESOLVER` and `ROLE_SET_TOKEN_OBSERVER` for the contract owner. We could do this by calling the relevant method after deployment, but perhaps it would be cleaner to add a `deployerRoles` argument to the constructor?

Arachnid commented on contracts/src/registry/PermissionedRegistry.sol:
```diff
@@ -2,55 +2,269 @@
+    {
+        tokenId = NameUtils.labelToTokenId(label);
+
+        (, uint64 oldExpiry, uint32 tokenIdVersion) = datastore.getSubregistry(tokenId);
```
Perhaps this should be combined with line `90` using a utility function that returns `(versionedTokenId, expiry)`?

Arachnid commented on contracts/src/registry/PermissionedRegistry.sol:
```diff
@@ -2,55 +2,269 @@
+        address previousOwner = super.ownerOf(tokenId);
+        if (previousOwner != address(0)) {
+            _burn(previousOwner, tokenId, 1);
+            tokenId = _regenerateTokenId(tokenId); // so we have a fresh acl
```
We already know the `registry` and `expires` here; can we pass them in to this function to avoid re-fetching them?

Arachnid commented on contracts/src/registry/PermissionedRegistry.sol:
```diff
@@ -2,55 +2,269 @@
+        onlyRootRoles(ROLE_REGISTRAR)
+        returns (uint256 tokenId)
+    {
+        tokenId = NameUtils.labelToTokenId(label);
```
I'm concerned about passing around multiple different things called `tokenId` that have different values - versioned and unversioned. Could we either rename unversioned uses, or avoid having this intermediate value in a variable?

hiddentao replied:
I'll use `tokenId` and `canonicalId` where the latter means with the bits zeroed out.

Arachnid commented on contracts/src/registry/RegistryDatastore.sol:
```diff
@@ -2,43 +2,47 @@
+import {NameUtils} from "../utils/NameUtils.sol";
 
 contract RegistryDatastore is IRegistryDatastore {
-    uint256 LABEL_HASH_MASK = 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffff00000000;
```
I think we still need to mask out the lower bits in the datastore, because that provides a guarantee to anyone reading the datastore storage about how token IDs will be treated.

hiddentao replied:
It still masks it out - I've just abstracted the logic into `NameUtils.getCanonicalId()`

Arachnid commented on contracts/src/registry/IStandardRegistry.sol:
```diff
@@ -0,0 +1,104 @@
+     */
+    function getExpiry(uint256 tokenId) external view returns (uint64);
+
+    /**
+     * @dev Fetches the resource of a subdomain.
+     * @param tokenId The token ID of the subdomain to fetch the resource for.
+     * @return The resource of the subdomain.
+     */
+    function getTokenIdResource(uint256 tokenId) external view returns(bytes32);
+
+    /**
+     * @dev Get the token ID of a resource.
+     * @param resource The resource to fetch the token ID for.
+     * @return The token ID of the resource.
+     */
+    function getResourceTokenId(bytes32 resource) external view returns (uint256);
+}
```
Do these need to be part of the public interface?

Arachnid commented on contracts/src/registry/PermissionedRegistry.sol:
```diff
@@ -2,55 +2,295 @@
+    {
+        uint64 oldExpiry;
+        uint32 tokenIdVersion;
+        (tokenId, oldExpiry, tokenIdVersion) = this.getNameData(label);
```
Calling via `this` makes it an external call, which is less efficient. Is this needed here?

Arachnid commented on contracts/src/registry/PermissionedRegistry.sol:
```diff
@@ -2,55 +2,295 @@
+
+        // if there is a previous owner, burn the token
+        address previousOwner = super.ownerOf(tokenId);
+        if (previousOwner != address(0)) {
+            _burn(previousOwner, tokenId, 1);
+            tokenId = _regenerateTokenId(tokenId, address(registry), expires, tokenIdVersion); // so we have a fresh acl
+        } else {
+            datastore.setSubregistry(tokenId, address(registry), expires, tokenIdVersion);
+        }
```
If you change `_regenerateTokenId` to take the new version instead of the old one, you can remove the `else` clause, increment the `tokenIdVersion` in the `if` clause, and call `_regenerateTokenId` outside the if clause to handle both cases. Maybe call it `generateTokenId` instead?

```suggestion
        if (previousOwner != address(0)) {
            _burn(previousOwner, tokenId, 1);
            tokenIdVersion++; // so we have a fresh acl
        }
        tokenId = _generateTokenId(tokenId, address(registry), expires, tokenIdVersion); 
```

Arachnid commented on contracts/src/registry/PermissionedRegistry.sol:
```diff
@@ -2,55 +2,295 @@
+
+        for (uint256 i = 0; i < ids.length; ++i) {
+            /*
+            in _regenerateToken, we burn the token and then mint a new one. This flow below ensures the roles go from owner => zeroAddr => owner
```
It's not clear to me what this comment means. Isn't the order the reverse of what's implied?

hiddentao replied:
the flow is:

1. burn
2. mint

in (1) the roles are copied from owner to zero address, in (2) they are copied from zero address back to the owner.

Arachnid commented on contracts/src/registry/PermissionedRegistry.sol:
```diff
@@ -2,55 +2,295 @@
+/**
+ * @dev Observer pattern for events on existing tokens.
+ */
+interface TokenObserver {
```
This should be in its own file, so that anyone wanting to implement an observer doesn't need to include the entire registry source code.

Arachnid commented on contracts/src/registry/ETHRegistrar.sol:
```diff
@@ -142,24 +142,20 @@ contract ETHRegistrar is IETHRegistrar, AccessControl {
 
-        tokenId = registry.register(name, owner, subregistry, resolver, flags, uint64(block.timestamp) + duration);
+        uint64 expiry = uint64(block.timestamp) + duration;
+        tokenId = registry.register(name, owner, subregistry, resolver, ALL_ROLES, expiry);
```
This needs adjusting - as it stands it will allow the user to renew their own name directly. Either this should mask out `ROLE_RENEW` and `ROLE_RENEW_ADMIN` or it should be set to only the safe roles.
